### PR TITLE
FINERACT-2111: Performance issue on Retrieve Loan API - Transactions

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

Describe the changes made and why they were made.

Exists a performance in the loans due to lots of transactions like repayments made on the loan, accruals and some of the real-time evaluations in below GET api uses loan.getTransactions and it takes more than 20seconds to return the response.

```
/fineract-provider/api/v1/loans/{loanId}
/fineract-provider/api/v1/loans/external-id/{externalloanId}

```
By knowing that we are introducing daily accruals in interest bearing product, its going to fail. Please rewrite the loading transactions by filter in the JPA query

[FINERACT-2111](https://issues.apache.org/jira/browse/FINERACT-2111)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java [hunk 1]
```diff
@@ -262,7 +266,8 @@ private CollectionData calculateDelinquencyDataForOverdueInstallment(final Loan
             final LoanRepaymentScheduleInstallment installment) {
         final MonetaryCurrency loanCurrency = loan.getCurrency();
         LoanRepaymentScheduleInstallment latestInstallment = loan.getLastLoanRepaymentScheduleInstallment();
-        List<LoanTransaction> chargebackTransactions = loan.getLoanTransactions(LoanTransaction::isChargeback);
+        Collection<LoanTransaction> chargebackTransactions = loanTransactionReadService.fetchLoanTransactionsByType(loan.getId(),
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java]: List as type would be better i think.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanTransactionReadService.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    private final EntityManager entityManager;
+
+    public Collection<LoanTransaction> fetchLoanTransactionsByType(final Long loanId, final Integer transactionType) {
+        final List<Integer> transactionTypes = new ArrayList<>();
+        transactionTypes.add(transactionType);
+        return fetchLoanTransactionsByTypes(loanId, transactionTypes);
+    }
+
+    public Collection<LoanTransaction> fetchLoanTransactionsByTypes(final Long loanId, final List<Integer> transactionTypes) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanTransactionReadService.java]: Make it a List pls

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 3]
```diff
@@ -287,10 +287,12 @@ public Collection<LoanTransactionData> retrieveLoanTransactions(final Long loanI
             // TODO: would worth to rework in the future. It is not nice to fetch relations one by one... might worth to
             // give a try to get rid of native queries
             for (LoanTransactionData loanTransaction : loanTransactionData) {
-                loanTransaction
-                        .setLoanTransactionRelations(this.retrieveLoanTransactionRelationsByLoanTransactionId(loanTransaction.getId()));
-                loanTransaction.setLoanChargePaidByList(
-                        loanChargePaidByReadPlatformService.getLoanChargesPaidByTransactionId(loanTransaction.getId()));
+                if (loanTransaction.supportTransactionRelations()) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: An idea:

1. Please create a list of loan transactions ids that were fetched in the loan transaction data.
2. Use this transaction id list to fetch all the loan transaction relations data
3. Use this transaction id list to fetch all the loan charge paid by data
4. Iterate througn on loan transaction data and set from the above list the transaction relations and charge paid by's if applicable

This way we will have only 3 queries!

**IMPORTANT!**
transaction id list can be quite huge.

Please fetch the value from `fineract.query.in-clause-parameter-size-limit` and use this to trigger smaller chunks to fetch the data. You can after just simply merge the lists together and do the above processing.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanTransactionReadService.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanTransactionReadService.java]: Done! updated

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/delinquency/service/LoanDelinquencyDomainServiceImpl.java]: Done! Updated

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 4]
```diff
@@ -286,9 +283,13 @@ public Collection<LoanTransactionData> retrieveLoanTransactions(final Long loanI
...
+            final List<Long> loanIds = loanTransactionData.stream().map(LoanTransactionData::getId).collect(Collectors.toList());
+            final List<LoanTransactionRelationData> loanTransactionRelationDatas = loanTransactionRelationReadService
+                    .fetchLoanTransactionRelationDataFrom(loanIds);
             for (LoanTransactionData loanTransaction : loanTransactionData) {
-                loanTransaction
-                        .setLoanTransactionRelations(this.retrieveLoanTransactionRelationsByLoanTransactionId(loanTransaction.getId()));
+                loanTransaction.setLoanTransactionRelations(loanTransactionRelationDatas.stream().filter(
+                        loanTransactionRelationData -> loanTransactionRelationData.getFromLoanTransaction().equals(loanTransaction.getId()))
+                        .toList());
                 loanTransaction.setLoanChargePaidByList(
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Please do the same optimization for the `loan charge paid by` as well.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 5]
```diff
@@ -1224,7 +1226,6 @@ public LoanScheduleData extractData(@NotNull final ResultSet rs) throws SQLExcep
                 final LocalDate dueDate = JdbcSupport.getLocalDate(rs, "dueDate");
                 final LocalDate obligationsMetOnDate = JdbcSupport.getLocalDate(rs, "obligationsMetOnDate");
                 final boolean complete = rs.getBoolean("complete");
-                final boolean isAdditional = rs.getBoolean("isAdditional");
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:1227]: Was this deliberate?

Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:1227]: Not really, It was not used

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:1227]: Nice catch in that  case! :)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Done! Updated

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**